go.mod

```diff
@@ -3,6 +3,7 @@ module github.com/canonical/lxd
 go 1.22.4
 
 require (
+	github.com/NVIDIA/nvidia-container-toolkit v1.15.0
```

Comment:
@elezar From a UX point of view, we'd prefer keeping the generation logic inside LXD to fully own and control it. Moreover, since LXD is deployed inside a snap, the root path for the discovery is likely to change according on how the LXD snap is provided with NVIDIA userspace libs. There can be two cases:

* The 'classic' case where the drivers and libs are already installed on the host (here we could perfectly let the user generate its own CDI spec)
* The 'snap isolated' case where the drivers and libs are brought by an other snap (we calll them provider snap). To do this 'provide' operation, we connect snaps together: this is done like `snap connect lxd:graphics-core22 nvidia-core22:graphics-core22` . This is a use case that will be needed for Ubuntu Core (basically, an ubuntu flavor where everythings are contained in snaps, even the kernel and managed by snapd)... In this case, we couldn't let a user generate the spec. 

In order to remain consistent with both cases, that is why we will let LXD generate the spec.

Reply:
OK. That's a fair point. This is definitely something that we should keep in mind when making changes to the CDI spec generation API. This would be the first direct external client of this code. We have been more willing to make API breaking changes up to now given that we are the only consumers of this.

With this in mind, are there any concerns about the "shape" of the API from your side?

Reply:
@dtrudg for info. We discussed something similar in the context of non-OCI singularity some time back. The changes here may be of interest to you for https://github.com/sylabs/singularity/issues/1395

Reply:
I don't think I have much concerns overall. The API client is well engineered and understandable. Maybe that the most important for us  would be to keep the functions instructing the CDI generator on 'where' to look for the discovery (the `nvcdi.WithDriverRoot` and `nvcdi.WiithDevRoot` are a real necessity for us and are as of now, cleverly thought :) ) I noticed that the CDI discovery was also trying to resolve the `x86_64-linux-gnu` triplet whereas the platform was arm64: maybe allowing the API to be 'arch ready' (maybe with a `WithArchTriplet(...)`) would be also a nice addition :)  

The other thing I had in the back of my mind but this would be just a bonus, would be to maybe have a `nvcdi.WithCapabilities` function to pass the [runtime capabilities](https://github.com/NVIDIA/nvidia-container-toolkit/tree/main/cmd/nvidia-container-runtime#nvidia_driver_capabilities) and generate only a subset of the libs and drivers to passthrough for achieving the capabilities. 

For,

Reply:
cc @tomponline

Reply:
You could override much of this with `WithLibrarySearchPaths(paths []string)` if you know where to look for the libraries. 

We are working with an internal team to improve the CDI spec generation using information directly from the driver. With this in place we should be able to provide more flexibility in terms of filtering by "capabilities", for example.

Reply:
Thanks for the pro tip @elezar I'll mark this as fixed then.